satpy/readers/seviri_l2_bufr.py

```diff
@@ -0,0 +1,225 @@
+                ec.codes_set(bufr, 'unpack', 1)
+                arr = ec.codes_get_array(bufr, parameter)
+                ec.codes_release(bufr)
+
+            else:
+                msgCount = 0
+                while True:
+
+                    bufr = ec.codes_bufr_new_from_file(fh)
+                    if bufr is None:
+                        break
+
+                    ec.codes_set(bufr, 'unpack', 1)
+                    # if is the first message initialise our final array with
+                    # the number of subsets contained in the first message
+                    if (msgCount == 0):
+                        arr = np.zeros(ec.codes_get(bufr, 'numberOfSubsets'), np.float)
+                        arr[:] = ec.codes_get_array(bufr, parameter, float)
+
+                    else:
+                        tmpArr = np.zeros(ec.codes_get(bufr, 'numberOfSubsets'), np.float)
+                        tmpArr[:] = ec.codes_get_array(bufr, parameter, float)
+                        arr = np.concatenate((arr, tmpArr))
+
+                    msgCount = msgCount+1
+                    ec.codes_release(bufr)
+
+        fh.close()
+        if arr.size == 1:
+            arr = arr[0]
+
+        return arr
+
+    def get_attribute(self, parameter, mnbr=None):
+        ''' Get BUFR attributes '''
+
+        fh = open(self.filename, "rb")
+
+        if mnbr is not None:
+            bufr = ec.codes_bufr_new_from_file(fh)
+            ec.codes_set(bufr, 'unpack', 1)
+            arr = ec.codes_get_array(bufr, parameter)
+            ec.codes_release(bufr)
+
+        fh.close()
+
+        if arr.size == 1:
+            arr = arr[0]
+
+        return arr
+
+    def get_dataset(self, dsid, info):
+        ''' here we loop through the BUFR file and for the required key
+        append the data vlues to the dataset area'''
+
+        arr2 = np.empty((self.Ny, self.Nx)).astype(np.float)
+        arr2.fill(np.nan)
+
+        with open(self.filename, "rb") as fh:
+            fh = open(self.filename, "rb")
+
+            parameter = info['key']
+
+            msgCount = 0
+            while True:
+
+                bufr = ec.codes_bufr_new_from_file(fh)
+                if bufr is None:
+                    break
+
+                ec.codes_set(bufr, 'unpack', 1)
+                # if is the first message initialise our final array with
+                # the number of subsets contained in the first message
+                if (msgCount == 0):
+                    arr = np.zeros(ec.codes_get(bufr, 'numberOfSubsets'), np.float)
+                    arr[:] = ec.codes_get_array(bufr, parameter, float)
+
+                else:
+                    tmpArr = np.zeros(ec.codes_get(bufr, 'numberOfSubsets'), np.float)
+                    tmpArr[:] = ec.codes_get_array(bufr, parameter, float)
+                    arr = np.concatenate((arr, tmpArr))
+
+                msgCount = msgCount+1
+                ec.codes_release(bufr)
+
+        fh.close()
```

Comment:
the `with` statement should take care of closing the file

Reply:
removed